### ISSUE TITLE
E-mails are not sent well

### ISSUE THREAD

**Issue by Miminoux** [glpi-project/glpi#1873]
Hi,

First, thx a lot for GLPI, just awesome...

I am meeting a strange behavior with GLPI : 
- I configured a notification on new reservations to notice the requester and a specific group
- I make a new reservation

Results :
- Sometimes, the requester receives the notification, sometimes not
- The specifiq group never receives the e-mail

Then I went into the "mail queue" menu and saw many of the e-mails never received. I first thought they had never been sent and went to check automatic action "queuedmail" : It was normaly executed (every minute) with the CLI method. From here, my head has been driving crazy :  how could the automatic action "queuedmail" be run well and e-mails been stuck in the mail queue ?
Then I thought the problem was arround queuedmail and sceptically clicked "Execute". Then I became completely mad because it worked like a charm, the mail queue got sent the next second.

I am not sure but it seems some messages were missing between the configured recipents on that specifiq notification and messages in the mail queue. I'm going to check that tomorrow and update the track but it is not the main problem.
The main issue for me the fact that some e-mails were blocked in queue while queuedmail action was running nicely.

I'am running GLPI 9.1.2 with some plugins : 
	news                 Name: Alertes                        Version: 1.3.2.5    State: Enabled
	badges               Name: Badges                         Version: 2.2.0      State: Enabled
	fields               Name: Champs supplémentaires         Version: 1.4.5      State: Enabled
	behaviors            Name: Comportements                  Version: 1.2        State: Enabled
	formcreator          Name: Formulaires                    Version: 2.4.0      State: Enabled
	fusioninventory      Name: FusionInventory                Version: 9.1+1.1    State: Enabled
	tag                  Name: Gestion des tags               Version: 0.90-1.3   State: Enabled
	datainjection        Name: Injection de fichiers          Version: 2.4.2      State: Enabled
	dashboard            Name: Tableau de bord                Version: 0.8.0      State: Enabled

Bye

**Comment by tomolimo** [comment]
Hello 
I confirm this strange behavior. I believe that it's coming from the crowd, that logs in the log files that there is nothing to launch, even if some automatic actions are waiting. 
Regards 
Tomolimo

**Comment by joraff** [comment]
We just discovered an issue related to this: the cron job was configured to send 50 emails at one time, however it was always pulling 50 emails with SMTP delivery errors. The `sent_try` field kept incrementing on this set indefinitely. Check those messages in your queue for this.

There's no current mechanism for a maximum retries. Project maintainers, I am happy to PR something to take care of this if we can agree on the appropriate way to implement.

First thought: add an `is_fatal` column to `glpi_queuedmails` if `sent_try` exceeds at set maximum number of retries. Cons: schema change, possible database bloat, notify admin?

**Comment by joraff** [comment]
Possible way around the schema change and db bloat: write the event to mail.log and mark as is_deleted ?

**Comment by joraff** [comment]
@trasher

**Comment by trasher** [comment]
I guess using `is_deleted` would be confusing; as it is apparently already used elsewhere.
I also guess we should rely on the `send_try` value only; no need to add any extra field.

+1 for an entry in a log if we do not already have something (I do not know the mail part at all).

Also, note that we could implement a send limit to the 9.1.x; but only if there are no changes on the database model. We can hardcore the value, sayt to 5 maximum retries; and display differently queued messages in the UI.
Maybe can we rely on a constant in `define.php` users may change if needed?

For 9.2 (current master); we can add a configuration for the max retries; and log information elsewhere so we can display what happened on the UI.

I wonder if we could use history (or maybe is it already used atsome point) in order to store what happended.

**Comment by joraff** [comment]
The normal life for a msg in the mail queue is to try and send, if sent: save the sent timestamp, mark as is_deleted and log the event in mail.log, if not sent: increment sent_try and move on. Eventually another cron job removes the is_deleted message from the queue.

I don't think anything should be added to define.php in 9.1 or 9.2 since that would change default behavior. Maybe 9.3. The PR above is backwards compatible with both branches and only changes mail queye behavior if a $CFG is set. And it only gets set if they visit the smtp config screen and allow max retries to be saved.

**Comment by trasher** [comment]
Hum OK, I understand; thanks for the details. Adding this feature for 9.2 is perfectly OK to me; if messages are not deleted from the queue (that way, default behavior will be changed; but not so much).

I don't think relying on the config to be defined is a good idea; we would have 2 differents uses cases depending on user configuration...

Maybe should we just set the default configuration to `-1` for example, to disable the max tries in a first time? If user want, he'll set the max tries in its configuration, and will know what would happen.

**Comment by joraff** [comment]
> Maybe should we just set the default configuration to -1 for example, to disable the max tries in a first time? If user want, he'll set the max tries in its configuration, and will know what would happen.

The end result you describe here is exactly what this PR does. The max tries does nothing unless the user types a value in the notification mail settings form and saves it (I removed the html placeholder value). Setting a default to -1 complicates what to display in the HTML `<input />` field. 

For 9.2 I can submit a different PR that adds a default value to define.php to enable the behavior without user input.

**Comment by dvinkic** [comment]
I have problem like Miminoux.
My version is 9.1 and the problem is:

Notification is set to cli for 1 minute and cron job is set to 1 minute

In logs of Automatic action - queuedmail I get:
26-02-2017 11:06    0.013 seconds       0    Action completed, no processing required
26-02-2017 11:05    0.012 seconds       0    Action completed, no processing required
26-02-2017 11:04    0.012 seconds       0    Action completed, no processing required

The job started but no mail from mail queue is sent.

Then I manualy Execute the action and in log I get:
26-02-2017 12:07    0.678 seconds       3    Action completed, fully processed

It is the same when I change from cli to glpi

When the ticket is closed the notification mail is sent
When ticket is updated (change urgency, add item, add resolve date) the mail is sent.

The problem is only when tasks and follow ups are created, then they go to mail queue.

And then they stay in the queue until I go to Automatic action - queuedmail and click to execute (they are not automaticly executed) there is only 3-4 mails to send

try with CLI and GLPI run mode.

**Comment by tomolimo** [comment]
Hello 
Could you please give complete copy of logs? 
Thank you 
Tomolimo

**Comment by tomolimo** [comment]
Hello, 
In case it may help you, I solved my problem: I was using opcache with option cli set to On. I set it to Off, and the issue disappeared. The emails are now sent on time. 
Regards 
Tomolimo

**Comment by dvinkic** [comment]
Hi Tomolimo,
Thanks for you replay.
opcache is disabled in php.ini

Here is the access log for glpi server:

First in the existing ticket I add a followup and there are 3 items in mail queue (I set 3 mail addresses to be notified)

104 - - [15/Mar/2017:22:03:16 +0100] "GET /front/queuedmail.php HTTP/1.1" 200 28018
104 - - [15/Mar/2017:22:03:16 +0100] "GET /front/cron.php HTTP/1.1" 200 63
104 - - [15/Mar/2017:22:03:29 +0100] "GET / HTTP/1.1" 200 2219
104 - - [15/Mar/2017:22:03:30 +0100] "GET /front/cron.php HTTP/1.1" 200 63
104 - - [15/Mar/2017:22:03:39 +0100] "POST /front/login.php HTTP/1.1" 200 7316
104 - - [15/Mar/2017:22:03:44 +0100] "GET /front/logout.php?noAUTO=1 HTTP/1.1" 200 269
104 - - [15/Mar/2017:22:03:44 +0100] "GET /index.php?noAUTO=1 HTTP/1.1" 200 2266
104 - - [15/Mar/2017:22:03:44 +0100] "GET /front/cron.php HTTP/1.1" 200 63
104 - - [15/Mar/2017:22:03:50 +0100] "GET /front/user.php HTTP/1.1" 200 50105
104 - - [15/Mar/2017:22:04:15 +0100] "POST /front/login.php HTTP/1.1" 200 297
104 - - [15/Mar/2017:22:04:16 +0100] "GET /front/helpdesk.public.php HTTP/1.1" 200 14300
104 - - [15/Mar/2017:22:04:21 +0100] "GET /front/ticket.php?criteria[0][field]=12&criteria[0][searchtype]=equals&criteria[0][value]=2&criteria[0][link]=AND&reset=reset HTTP/1.1" 200 32037
104 - - [15/Mar/2017:22:04:30 +0100] "GET /front/ticket.form.php?id=6 HTTP/1.1" 200 15947
104 - - [15/Mar/2017:22:04:30 +0100] "GET /ajax/common.tabs.php?_target=/front/ticket.form.php&_itemtype=Ticket&_glpi_tab=Ticket$main&id=6& HTTP/1.1" 200 12586
104 - - [15/Mar/2017:22:04:32 +0100] "GET /ajax/common.tabs.php?_target=/front/ticket.form.php&_itemtype=Ticket&_glpi_tab=Ticket$1&id=6& HTTP/1.1" 200 6634
104 - - [15/Mar/2017:22:04:33 +0100] "GET /pics/timeline/todo.png HTTP/1.1" 200 344
104 - - [15/Mar/2017:22:04:33 +0100] "GET /pics/timeline/task.png HTTP/1.1" 200 1301
104 - - [15/Mar/2017:22:04:33 +0100] "GET /pics/timeline/actiontime.png HTTP/1.1" 200 478
104 - - [15/Mar/2017:22:04:32 +0100] "GET /ajax/updatecurrenttab.php?itemtype=Ticket&id=6&tab=1 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:04:33 +0100] "GET /front/document.send.php?file=_pictures/11/7_58a96eb829511_min.jpg HTTP/1.1" 200 2992
104 - - [15/Mar/2017:22:04:40 +0100] "POST /ajax/timeline.php HTTP/1.1" 200 6909
104 - - [15/Mar/2017:22:05:08 +0100] "POST /front/ticketfollowup.form.php HTTP/1.1" 200 337
104 - - [15/Mar/2017:22:05:09 +0100] "GET /front/ticket.form.php?id=6 HTTP/1.1" 200 15962
104 - - [15/Mar/2017:22:05:09 +0100] "GET /ajax/common.tabs.php?_target=/front/ticket.form.php&_itemtype=Ticket&_glpi_tab=Ticket$1&id=6& HTTP/1.1" 200 7717
104 - - [15/Mar/2017:22:05:10 +0100] "GET /front/document.send.php?file=_pictures/11/7_58a96eb829511_min.jpg HTTP/1.1" 200 2992
104 - - [15/Mar/2017:22:05:10 +0100] "GET /pics/timeline/followup.png HTTP/1.1" 200 426
104 - - [15/Mar/2017:22:05:23 +0100] "GET /front/queuedmail.php HTTP/1.1" 200 42206
104 - - [15/Mar/2017:22:05:33 +0100] "GET /front/crontask.php HTTP/1.1" 200 57064
104 - - [15/Mar/2017:22:05:41 +0100] "GET /front/crontask.form.php?id=22 HTTP/1.1" 200 25239
104 - - [15/Mar/2017:22:05:41 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTaskLog$2&id=22& HTTP/1.1" 200 17096
104 - - [15/Mar/2017:22:05:47 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTask$main&id=22& HTTP/1.1" 200 19787
104 - - [15/Mar/2017:22:05:47 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=0 HTTP/1.1" 200 -


The cron is working but the items are still in the queue.

Then I create new ticket

104 - - [15/Mar/2017:22:08:59 +0100] "GET /front/helpdesk.public.php HTTP/1.1" 200 14300
104 - - [15/Mar/2017:22:09:12 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=2 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:09:21 +0100] "GET /front/queuedmail.php HTTP/1.1" 200 42266
104 - - [15/Mar/2017:22:09:22 +0100] "GET /front/cron.php HTTP/1.1" 200 63
104 - - [15/Mar/2017:22:10:07 +0100] "GET /front/helpdesk.public.php?create_ticket=1 HTTP/1.1" 200 40857
104 - - [15/Mar/2017:22:10:08 +0100] "GET /front/cron.php HTTP/1.1" 200 63
104 - - [15/Mar/2017:22:10:08 +0100] "POST /ajax/uemailUpdate.php HTTP/1.1" 200 1298
104 - - [15/Mar/2017:22:10:11 +0100] "POST /ajax/getDropdownValue.php HTTP/1.1" 200 559
104 - - [15/Mar/2017:22:10:13 +0100] "POST /ajax/comments.php HTTP/1.1" 200 91
104 - - [15/Mar/2017:22:10:13 +0100] "POST /front/tracking.injector.php HTTP/1.1" 200 41400
104 - - [15/Mar/2017:22:10:14 +0100] "POST /ajax/uemailUpdate.php HTTP/1.1" 200 1298
104 - - [15/Mar/2017:22:10:30 +0100] "POST /front/tracking.injector.php HTTP/1.1" 200 13606
104 - - [15/Mar/2017:22:10:45 +0100] "GET /front/crontask.php HTTP/1.1" 200 57058
104 - - [15/Mar/2017:22:10:56 +0100] "GET /front/crontask.form.php?id=22 HTTP/1.1" 200 25239
104 - - [15/Mar/2017:22:10:57 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTaskLog$2&id=22& HTTP/1.1" 200 17099
104 - - [15/Mar/2017:22:11:14 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTask$main&id=22& HTTP/1.1" 200 19808
104 - - [15/Mar/2017:22:11:14 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=0 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:11:15 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTaskLog$1&id=22& HTTP/1.1" 200 996
104 - - [15/Mar/2017:22:11:15 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=1 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:11:18 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=0 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:11:21 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=2 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:11:24 +0100] "GET /front/queuedmail.php HTTP/1.1" 200 42212

And I get notification on all 3 email addreses but followup is still in the mail queue.

Then I go to Automatic action - queuedmail and click to Execute

104 - - [15/Mar/2017:22:17:56 +0100] "GET /front/crontask.php HTTP/1.1" 200 57130
104 - - [15/Mar/2017:22:17:56 +0100] "GET /front/cron.php HTTP/1.1" 200 63
104 - - [15/Mar/2017:22:18:02 +0100] "GET /front/crontask.form.php?id=22 HTTP/1.1" 200 25239
104 - - [15/Mar/2017:22:18:03 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTaskLog$2&id=22& HTTP/1.1" 200 17104
104 - - [15/Mar/2017:22:18:05 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTask$main&id=22& HTTP/1.1" 200 19802
104 - - [15/Mar/2017:22:18:05 +0100] "GET /ajax/updatecurrenttab.php?itemtype=CronTask&id=22&tab=0 HTTP/1.1" 200 -
104 - - [15/Mar/2017:22:18:07 +0100] "POST /front/crontask.form.php HTTP/1.1" 200 343
104 - - [15/Mar/2017:22:18:08 +0100] "GET /front/crontask.form.php?id=22 HTTP/1.1" 200 25209
104 - - [15/Mar/2017:22:18:09 +0100] "GET /ajax/common.tabs.php?_target=/front/crontask.form.php&_itemtype=CronTask&_glpi_tab=CronTask$main&id=22& HTTP/1.1" 200 19781
104 - - [15/Mar/2017:22:18:12 +0100] "GET /front/queuedmail.php HTTP/1.1" 200 27953

And finaly I got the mail notification for followup

Thank you for your help.
dvinkic

**Comment by tomolimo** [comment]
Thank you for your answer, but what I would like are the glpi logs and not the apache logs. And especially the GLPI cron logs. 
Thank you 
Tomolimo

**Comment by dvinkic** [comment]
GLPI cron logs:

External #1: Launch queuedmail
2017-03-15 21:59:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:00:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:01:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:02:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:03:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 23:04:29 [7@-----------]
Internal #1: Launch alertnotclosed
2017-03-15 22:04:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:05:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:06:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:07:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:08:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:09:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:10:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:11:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:12:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:13:55 [@-----------]
External #1: Launch queuedmail
2017-03-15 22:14:55 [@-----------]
External #1: Launch queuedmail

GLPI mail logs:

2017-03-15 22:05:08 [9@-----------]
An email to glpi.tehnicar@----------- was added to queue: [GLPI #0000006] New followup Create new instance
2017-03-15 22:05:09 [9@-----------]
An email to glpi.user@----------- was added to queue: [GLPI #0000006] New followup Create new instance
2017-03-15 22:05:09 [9@-----------]
An email to helpdesk@----------- was added to queue: [GLPI #0000006] New followup Create new instance
2017-03-15 22:10:30 [9@-----------]
An email to glpi.user@----------- was added to queue: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:31 [9@-----------]
An email to glpi.tehnicar@----------- was added to queue: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:31 [9@-----------]
An email to -------@----------- was added to queue: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:31 [9@-----------]
An email to -------@----------- was added to queue: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:31 [9@-----------]
An email was sent to glpi.user@-----------: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:31 [9@-----------]
An email was sent to glpi.tehnicar@-----------: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:31 [9@-----------]
An email was sent to -------@-----------: [GLPI #0000008] New ticket New ticket
2017-03-15 22:10:32 [9@-----------]
An email was sent to -------@-----------: [GLPI #0000008] New ticket New ticket
2017-03-15 22:18:08 [7@-----------]
An email was sent to glpi.tehnicar@-----------: [GLPI #0000006] New followup Create new instance
2017-03-15 22:18:08 [7@-----------]
An email was sent to glpi.user@-----------: [GLPI #0000006] New followup Create new instance
2017-03-15 22:18:08 [7@-----------]
An email was sent to helpdesk@-----------: [GLPI #0000006] New followup Create new instance

There are no php-errors
in event.log there is only login information.


Thank you for your help.
dvinkic

**Comment by tomolimo** [comment]
Thank you 
Could you please also post your glpi  cron configurations? And also the command line that starts the external cron? 
I don't remember if you already have posted the glpi configuration, if not, please post it also. 
Thank you 
Tomolimo

**Comment by dvinkic** [comment]
The line that starts external cron:
@ 1 /usr/bin/php /html/front/cron.php &>/dev/null

And the GLPI internal Automatic action - queuedmail
Name	queuedmail	Comments	
Description	Send mails in queue
Run frequency	1 minute
Status	Scheduled
Run mode CLI	
Run period	0 -24
 -> 
Number of days this action logs are stored	
	Last run	16-03-2017 11:19 Blank
Maximum emails to send at once 	
	Next run	As soon as possible.

Thank you for your help.
dvinkic

**Comment by tomolimo** [comment]
Hello
Thank you,
What is the value of "Maximum emails to send at once"?

**Comment by dvinkic** [comment]
Hi,
It is 50
No, thank you for your help.

**Comment by dvinkic** [comment]
Hi,
I think I found the problem, now to find why and solution.

Time is the problem, cron tasks are 1 hour behind all other automatic actions. When I wait for an hour the mails from queue are sent.

All other automatic actions working like a clock.

**Comment by dvinkic** [comment]
Found solution.
GLPI is working under in our case php-fpm and

When I set in automatic actions  - Run mode CLI - the global php.ini is used not from php-fpm (for every virtual server in our case)

Now the global php.ini timezone is change, this is the problem for other virtual servers. And GLPI cron is working and mails from queue are sent. 

The other solution is to connect cron.php through GLPI API and not to use separately. 

I forget to mention when Run mode is set to GLPI the mail is not sent from the mail queue, and because of that I use Run mode CLI, nor the mailgate from automatic actions is working. But only when working with mail is the case, other Run modes with glpi are working.
